src/frontend/src/expr/pure.rs

```diff
@@ -0,0 +1,43 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::{ExprImpl, ExprVisitor};
+
+struct ImpureAnalyzer {}
+
+impl ExprVisitor<bool> for ImpureAnalyzer {
+    fn merge(a: bool, b: bool) -> bool {
+        // the expr will be impure if any of its input is impure
+        a || b
+    }
+
+    fn visit_user_defined_function(&mut self, _func_call: &super::UserDefinedFunction) -> bool {
+        true
+    }
+
+    fn visit_function_call(&mut self, func_call: &super::FunctionCall) -> bool {
+        matches!(
+            func_call.get_expr_type(),
+            risingwave_pb::expr::expr_node::Type::Now | risingwave_pb::expr::expr_node::Type::Udf
+        )
```

Comment:
https://github.com/risingwavelabs/risingwave/blob/6791cb1e7a57b42f65ae1a8891c1bbfd5703630f/proto/expr.proto#L160-L168

Reply:
Also, the `Type::Udf` is unreachable here. 

Maybe we should add a `fn expr_type(&self)`  to each expr impl. And implement `visit_expr` (by `expr_type > 1000`) for `ImpureAnalyzer`

Reply:
This reminds me an issue that should we merge `UserDefinedFunction` into `FunctionCall`? 🤔

Reply:
`matches` or `_` can be error-prone. +1 on `expr_type`. 🤔

Reply:
No need to match all and use `1 < expr_node as i32 <= 600` instead?

Also PTAL at `FunctionCall::is_pure`.

Reply:
:innocent: is it a good engineering practice? When I am trying to add a new expression I can not know there is a assumption about the magic number(600) of the enum value

Reply:
So it seems your approach are better. good boy

Reply:
unified the pure check logic.